File: src/scripts/_lib/installed_tools.py
"""Project-scope installed-tools manifest at ``agents/installed-tools.lock``.

Phase 3 of road-to-global-first-install (ADR-008). Committed
bill-of-materials for AI tooling a project depends on. Sibling to the
global lockfile (``installed_lock.py``) but architecturally distinct:

- ``installed_lock.py`` lives in ``~/.event4u/agent-config/`` and tracks
  the user-scope environment (a single ``agent_config_version`` and a
  flat ``tools[]`` list).
- ``installed_tools.py`` lives in ``agents/`` and tracks **per-project**
  tooling with richer per-entry metadata (``scope``, ``bridge_marker``,
  ``installed_at``).

The file is machine-managed: ``init`` appends / merges; ``sync`` replays;
``validate`` drift-checks. Schema is YAML; ``pyyaml`` is used when
available, otherwise a constrained manual parser handles the documented
schema (no anchors, no flow style, single-level nesting under
``tools``).
"""
from __future__ import annotations

import os
import re
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Optional

try:
    from scripts._lib.fs_atomic import write_atomic  # noqa: PLC0415
except ImportError:  # pragma: no cover — alt sys.path layout (tests)
    from _lib.fs_atomic import write_atomic  # type: ignore[no-redef]  # noqa: PLC0415

MANIFEST_ENV = "AGENT_CONFIG_INSTALLED_TOOLS"
DEFAULT_MANIFEST_RELATIVE = Path("agents") / "installed-tools.lock"
SCHEMA_VERSION = 2

#: Schema versions older writers may have emitted. Reading any of these
#: succeeds; writing always produces :data:`SCHEMA_VERSION`.
SCHEMA_VERSIONS_SUPPORTED = (1, 2)

_VALID_SCOPES = ("global", "project")

#: Permitted values for ``files[].kind`` (P1.1, road-to-multi-package-
#: coexistence). ``bridge`` = team-pointer marker (e.g. ``.cursorrules``);
#: ``deployed`` = bundle content we wrote (e.g. ``.augment/rules/*.md``);
#: ``marker`` = one-off sentinel (e.g. ``claude-desktop`` install marker).
FILE_KINDS = frozenset({"bridge", "deployed", "marker"})

#: Stable known deploy roots — directories under which the doctor
#: command surveys for foreign files. Writers may extend the live
#: ``deploy_roots`` field per project; this constant is the canonical
#: default the installer seeds.
DEFAULT_DEPLOY_ROOTS = (
    ".augment/rules",
    ".augment/skills",
    ".augment/commands",
    ".cursor/rules",
    ".claude/skills",
    ".claude/commands",
    ".clinerules",
    ".windsurf/rules",
)


def manifest_path(project_root: Path, env: Optional[dict] = None) -> Path:
    """Return the active manifest path, honoring the env override."""
    env = env if env is not None else os.environ
    override = env.get(MANIFEST_ENV)
    if override:
        return Path(override).expanduser()
    return project_root / DEFAULT_MANIFEST_RELATIVE


# ---------------------------------------------------------------------------
# Read
# ---------------------------------------------------------------------------

_TOP_KEY_RE = re.compile(r'^([A-Za-z_][A-Za-z0-9_]*)\s*:\s*"?([^"\n]*?)"?\s*$')
_LIST_DASH_RE = re.compile(r"^\s*-\s*(.+?)\s*$")
_INDENT_KEY_RE = re.compile(r'^\s+([A-Za-z_][A-Za-z0-9_]*)\s*:\s*"?([^"\n]*?)"?\s*$')


def read_manifest(path: Path) -> Optional[dict[str, Any]]:
    """Parse the manifest into a dict; return ``None`` if absent.

    Tolerates partial / malformed files: missing keys yield missing dict
    entries rather than raising, so a corrupted file does not brick
    ``init``. v1 and v2 wire formats both return the same shape — v2
    optional fields (``deploy_roots``, per-tool ``files`` /
    ``merged_keys``) default to empty lists when absent, so callers can
    iterate without ``.get(..., [])`` boilerplate (P1.2).
    """
    try:
        text = path.read_text(encoding="utf-8")
    except (FileNotFoundError, OSError):
        return None
    data: Optional[dict[str, Any]] = None
    try:
        import yaml  # type: ignore[import-untyped]
        loaded = yaml.safe_load(text) or {}
        if isinstance(loaded, dict):
            data = loaded
    except ImportError:
        pass
    except Exception:
        # Fall through to the manual parser; corrupt YAML is recoverable
        # from our strict schema as long as the top-level shape holds.
        pass
    if data is None:
        data = _parse_manual(text)
    return _normalise_v2_shape(data)


def _normalise_v2_shape(data: dict[str, Any]) -> dict[str, Any]:
    """Backfill v2 optional fields so consumers can iterate uniformly.

    Idempotent: calling on an already-normalised dict is a no-op. Does
    not mutate input lists — replaces missing keys with fresh empties.
    """
    if data.get("tools") is None:
        data["tools"] = []
    if data.get("deploy_roots") is None:
        data["deploy_roots"] = []
    for tool in data["tools"]:
        if not isinstance(tool, dict):
            continue
        if tool.get("files") is None:
            tool["files"] = []
        if tool.get("merged_keys") is None:
            tool["merged_keys"] = []
    return data


def _parse_manual(text: str) -> dict[str, Any]:
    """Strict v1 manual parser; v2 nested fields are skipped, not raised.

    The fallback parser handles the canonical v1 wire format (top-level
    scalars + ``tools`` array with single-level key:value entries). For
    v2 manifests it still extracts the top-level scalars and the
    per-tool scalar fields, but silently drops nested arrays
    (``files``, ``merged_keys``) and top-level ``deploy_roots``. Callers
    that need full v2 fidelity must have ``pyyaml`` available — the
    manual path is a degraded read, not a v2 round-trip.
    """
    data: dict[str, Any] = {"tools": []}
    in_tools = False
    current: Optional[dict[str, Any]] = None
    # When the current tool entry opened a nested array (``files:`` or
    # ``merged_keys:``), we suppress recognition of the deeper ``- key``
    # lines as new tools until the indent climbs back to the per-tool
    # level (4 spaces).
    skip_until_outdent = False
    for raw in text.splitlines():
        stripped = raw.strip()
        if not stripped or stripped.startswith("#"):
            continue
        if stripped == "tools:":
            in_tools = True
            current = None
            skip_until_outdent = False
            continue
        if in_tools:
            indent = len(raw) - len(raw.lstrip(" "))
            if skip_until_outdent and indent > 4:
                continue
            skip_until_outdent = False
            m = _LIST_DASH_RE.match(raw)
            if m and indent == 2:
                first = m.group(1)
                current = {}
                data["tools"].append(current)
                inline = _TOP_KEY_RE.match(first)
                if inline:
                    current[inline.group(1)] = inline.group(2)
                continue
            mk = _INDENT_KEY_RE.match(raw)
            if mk and current is not None and indent == 4:
                key, val = mk.group(1), mk.group(2)
                if key in ("files", "merged_keys") and not val:
                    skip_until_outdent = True
                    continue
                current[key] = val
                continue
        m_top = _TOP_KEY_RE.match(raw)
        if m_top:
            key, value = m_top.group(1), m_top.group(2)
            if key == "deploy_roots" and not value:
                # Top-level v2 array — skip until next top-level scalar.
                in_tools = False
                current = None
                skip_until_outdent = True
                continue
            if key == "schema_version":
                try:
                    data[key] = int(value)
                except ValueError:
                    data[key] = value
            else:
                data[key] = value
            in_tools = False
            current = None
            skip_until_outdent = False
    return data


# ---------------------------------------------------------------------------
# Write
# ---------------------------------------------------------------------------


def _render(
    version: str,
    tools: list[dict[str, Any]],
    *,
    deploy_roots: Optional[list[str]] = None,
) -> str:
    lines = [
        f"schema_version: {SCHEMA_VERSION}",
        f'agent_config_version: "{version}"',
    ]
    if deploy_roots:
        lines.append("deploy_roots:")
        for root in deploy_roots:
            lines.append(f"  - {root}")
    lines.append("tools:")
    for tool in tools:
        lines.append(f"  - name: {tool['name']}")
        lines.append(f"    scope: {tool['scope']}")
        lines.append(f"    bridge_marker: {tool['bridge_marker']}")
        lines.append(f'    installed_at: "{tool["installed_at"]}"')
        status = tool.get("status")
        if status:
            lines.append(f"    status: {status}")
        files = tool.get("files") or []
        if files:
            # Sort by path ascending — deterministic output for golden-
            # file tests and stable team diffs (P1.3).
            files = sorted(files, key=lambda f: f["path"])
            lines.append("    files:")
            for entry in files:
                lines.append(f"      - path: {entry['path']}")
                lines.append(f"        kind: {entry['kind']}")
                sha = entry.get("sha256")
                if sha is None:
                    lines.append("        sha256: null")
                else:
                    lines.append(f'        sha256: "{sha}"')
        merged = tool.get("merged_keys") or []
        if merged:
            # Sort by (file, json_pointer) ascending — deterministic
            # output regardless of insertion order (P1.3).
            merged = sorted(
                merged, key=lambda e: (e["file"], e["json_pointer"]),
            )
            lines.append("    merged_keys:")
            for entry in merged:
                lines.append(f"      - file: {entry['file']}")
                lines.append(f"        json_pointer: \"{entry['json_pointer']}\"")
                vh = entry.get("value_hash")
                if vh is not None:
                    lines.append(f'        value_hash: "{vh}"')
    return "\n".join(lines) + "\n"


def write_manifest(
    path: Path,
    version: str,
    tools: list[dict[str, Any]],
    *,
    deploy_roots: Optional[list[str]] = None,
) -> Path:
    """Atomically write the manifest; return the path written.

    Delegates to :func:`scripts._lib.fs_atomic.write_atomic` so the
    crash-safety guarantees (fsync file, atomic rename, fsync parent
    dir) are shared with every other v2 writer. See P1.0 of
    ``road-to-multi-package-coexistence`` for the rationale.

    ``deploy_roots`` is the optional top-level v2 field listing
    directories the doctor command surveys for foreign files. When
    omitted, the field is not emitted (callers may rely on
    :data:`DEFAULT_DEPLOY_ROOTS` for the survey scope).
    """
    rendered = _render(version, tools, deploy_roots=deploy_roots)
    return write_atomic(path, rendered)


# ---------------------------------------------------------------------------
# Mutation helpers
# ---------------------------------------------------------------------------


class ScopeMismatchError(RuntimeError):
    """Raised when an existing manifest entry conflicts with the new scope."""

    def __init__(self, name: str, recorded_scope: str, new_scope: str):
        super().__init__(
            f"tool {name!r} is committed as scope={recorded_scope}; "
            f"refusing to change it to scope={new_scope} without --force"
        )
        self.name = name
        self.recorded_scope = recorded_scope
        self.new_scope = new_scope


def upsert_tool(
    existing: list[dict[str, Any]],
    *,
    name: str,
    scope: str,
    bridge_marker: str,
    installed_at: Optional[str] = None,
    force: bool = False,
    files: Optional[list[dict[str, Any]]] = None,
    merged_keys: Optional[list[dict[str, Any]]] = None,
) -> list[dict[str, Any]]:
    """Return a new tools list with ``name`` added or refreshed.

    Idempotency rules from ADR-008 §Lifecycle:
    * Same name, same scope → no-op (timestamp preserved).
    * Same name, different scope → raise ``ScopeMismatchError`` unless
      ``force=True``, in which case the entry is rewritten.
    * New name → appended in install order (not alphabetised).

    ``files`` / ``merged_keys`` are the v2 per-tool inventories
    (P1.4). When provided, they replace whatever was previously
    recorded on the entry — the installer is authoritative for the
    set of artefacts it just wrote. When ``None``, existing values
    are preserved on the idempotent path and absent on first-write.
    """
    if scope not in _VALID_SCOPES:
        raise ValueError(f"scope must be one of {_VALID_SCOPES}: {scope!r}")
    stamp = installed_at or _today()

    def _build(prior: Optional[dict[str, Any]] = None) -> dict[str, Any]:
        entry: dict[str, Any] = {
            "name": name,
            "scope": scope,
            "bridge_marker": bridge_marker,
            "installed_at": stamp,
        }
        if files is not None:
            entry["files"] = list(files)
        elif prior is not None and prior.get("files"):
            entry["files"] = list(prior["files"])
        if merged_keys is not None:
            entry["merged_keys"] = list(merged_keys)
        elif prior is not None and prior.get("merged_keys"):
            entry["merged_keys"] = list(prior["merged_keys"])
        return entry

    result: list[dict[str, Any]] = []
    found = False
    for entry in existing:
        if entry.get("name") == name:
            found = True
            recorded = str(entry.get("scope", ""))
            if recorded == scope:
                if files is None and merged_keys is None:
                    # Idempotent no-op — preserve original entry.
                    result.append(entry)
                else:
                    # Refresh inventories, preserve installed_at.
                    refreshed = _build(prior=entry)
                    refreshed["installed_at"] = entry.get(
                        "installed_at", stamp,
                    )
                    result.append(refreshed)
                continue
            if not force:
                raise ScopeMismatchError(name, recorded, scope)
            result.append(_build(prior=entry))
            continue
        result.append(entry)
    if not found:
        result.append(_build())
    return result


def _today() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%d")

File: src/scripts/_lib/installed_tools.ts
/**
 * Project-scope installed-tools manifest at `agents/installed-tools.lock`.
 *
 * TypeScript twin of `src/scripts/_lib/installed_tools.py` (ADR-200 — Python→TS
 * migration, Phase 2 / Wave 2a). Public API mirrors the Python module exactly
 * (snake_case kept deliberately — fidelity over TS idiom), including the
 * byte-exact YAML wire format the `_render` golden tests pin.
 *
 * Phase 3 of road-to-global-first-install (ADR-008). Committed
 * bill-of-materials for AI tooling a project depends on. Sibling to the
 * global lockfile (`installed_lock.ts`) but architecturally distinct:
 *
 * - `installed_lock` lives in `~/.event4u/agent-config/` and tracks the
 *   user-scope environment (a single `agent_config_version` and a flat
 *   `tools[]` list).
 * - `installed_tools` lives in `agents/` and tracks **per-project** tooling
 *   with richer per-entry metadata (`scope`, `bridge_marker`, `installed_at`).
 *
 * The file is machine-managed: `init` appends / merges; `sync` replays;
 * `validate` drift-checks. Schema is YAML; the Python reference uses `pyyaml`
 * when available and otherwise a constrained manual parser. This port has no
 * YAML library dependency — it always uses the manual parser, which handles
 * the documented schema (no anchors, no flow style, single-level nesting under
 * `tools`).
 *
 * Divergence candidate (flagged, not fixed): the Python reader prefers
 * `pyyaml` and only falls back to the manual parser when it is absent or the
 * YAML is malformed. This port has no YAML dependency, so it always takes the
 * manual-parser path. For the canonical schema both paths agree; for v2
 * manifests with nested `files` / `merged_keys` arrays the manual path
 * silently drops those nested fields (documented degraded read in the Python
 * source). A consumer relying on `pyyaml`-backed full-v2 fidelity would see a
 * behavior change here. See ADR-200 § intentional-divergence.
 */
import * as fs from "node:fs";
import * as os from "node:os";
import * as fsPath from "node:path";

import { write_atomic } from "./fs_atomic.js";

export type EnvMap = Record<string, string | undefined>;

/** A loosely-typed dict, mirroring Python's `dict[str, Any]`. */
export type AnyDict = Record<string, unknown>;

export const MANIFEST_ENV = "AGENT_CONFIG_INSTALLED_TOOLS";
export const DEFAULT_MANIFEST_RELATIVE = fsPath.join("agents", "installed-tools.lock");
export const SCHEMA_VERSION = 2;

/**
 * Schema versions older writers may have emitted. Reading any of these
 * succeeds; writing always produces {@link SCHEMA_VERSION}.
 */
export const SCHEMA_VERSIONS_SUPPORTED: readonly number[] = [1, 2];

const _VALID_SCOPES: readonly string[] = ["global", "project"];

/**
 * Permitted values for `files[].kind` (P1.1, road-to-multi-package-
 * coexistence). `bridge` = team-pointer marker (e.g. `.cursorrules`);
 * `deployed` = bundle content we wrote (e.g. `.augment/rules/*.md`);
 * `marker` = one-off sentinel (e.g. `claude-desktop` install marker).
 */
export const FILE_KINDS: ReadonlySet<string> = new Set(["bridge", "deployed", "marker"]);

/**
 * Stable known deploy roots — directories under which the doctor command
 * surveys for foreign files. Writers may extend the live `deploy_roots` field
 * per project; this constant is the canonical default the installer seeds.
 */
export const DEFAULT_DEPLOY_ROOTS: readonly string[] = [
  ".augment/rules",
  ".augment/skills",
  ".augment/commands",
  ".cursor/rules",
  ".claude/skills",
  ".claude/commands",
  ".clinerules",
  ".windsurf/rules",
];

/** Expand a leading `~` like Python's `Path.expanduser()`. */
function expanduser(p: string): string {
  if (p === "~") {
    return os.homedir();
  }
  if (p.startsWith("~/") || (process.platform === "win32" && p.startsWith("~\\"))) {
    return fsPath.join(os.homedir(), p.slice(2));
  }
  return p;
}

/** Return the active manifest path, honoring the env override. */
export function manifest_path(project_root: string, env?: EnvMap | null): string {
  const env_map = env ?? process.env;
  const override = env_map[MANIFEST_ENV];
  if (override) {
    return expanduser(override);
  }
  return fsPath.join(project_root, DEFAULT_MANIFEST_RELATIVE);
}

// ---------------------------------------------------------------------------
// Read
// ---------------------------------------------------------------------------

const _TOP_KEY_RE = /^([A-Za-z_][A-Za-z0-9_]*)\s*:\s*"?([^"\n]*?)"?\s*$/;
const _LIST_DASH_RE = /^\s*-\s*(.+?)\s*$/;
const _INDENT_KEY_RE = /^\s+([A-Za-z_][A-Za-z0-9_]*)\s*:\s*"?([^"\n]*?)"?\s*$/;

/**
 * Parse the manifest into a dict; return `null` if absent.
 *
 * Tolerates partial / malformed files: missing keys yield missing dict entries
 * rather than throwing, so a corrupted file does not brick `init`. v1 and v2
 * wire formats both return the same shape — v2 optional fields (`deploy_roots`,
 * per-tool `files` / `merged_keys`) default to empty lists when absent, so
 * callers can iterate without `?? []` boilerplate (P1.2).
 */
export function read_manifest(path: string): AnyDict | null {
  let text: string;
  try {
    text = require_read_text(path);
  } catch {
    return null;
  }
  // This port has no YAML dependency — always use the manual parser. See the
  // module-level divergence note.
  const data = _parse_manual(text);
  return _normalise_v2_shape(data);
}

/** Read a file as UTF-8 text, throwing on any error (caller catches). */
function require_read_text(path: string): string {
  return fs.readFileSync(path, { encoding: "utf-8" });
}

/**
 * Backfill v2 optional fields so consumers can iterate uniformly.
 *
 * Idempotent: calling on an already-normalised dict is a no-op. Does not
 * mutate input lists — replaces missing keys with fresh empties.
 */
export function _normalise_v2_shape(data: AnyDict): AnyDict {
  if (data["tools"] === undefined || data["tools"] === null) {
    data["tools"] = [];
  }
  if (data["deploy_roots"] === undefined || data["deploy_roots"] === null) {
    data["deploy_roots"] = [];
  }
  const tools = data["tools"] as unknown[];
  for (const tool of tools) {
    if (typeof tool !== "object" || tool === null) {
      continue;
    }
    const t = tool as AnyDict;
    if (t["files"] === undefined || t["files"] === null) {
      t["files"] = [];
    }
    if (t["merged_keys"] === undefined || t["merged_keys"] === null) {
      t["merged_keys"] = [];
    }
  }
  return data;
}

/**
 * Strict v1 manual parser; v2 nested fields are skipped, not raised.
 *
 * Handles the canonical v1 wire format (top-level scalars + `tools` array with
 * single-level key:value entries). For v2 manifests it still extracts the
 * top-level scalars and the per-tool scalar fields, but silently drops nested
 * arrays (`files`, `merged_keys`) and top-level `deploy_roots`.
 */
export function _parse_manual(text: string): AnyDict {
  const data: AnyDict = { tools: [] };
  const tools = data["tools"] as AnyDict[];
  let in_tools = false;
  let current: AnyDict | null = null;
  // When the current tool entry opened a nested array (`files:` or
  // `merged_keys:`), suppress recognition of the deeper `- key` lines as new
  // tools until the indent climbs back to the per-tool level (4 spaces).
  let skip_until_outdent = false;
  for (const raw of splitlines(text)) {
    const stripped = raw.trim();
    if (!stripped || stripped.startsWith("#")) {
      continue;
    }
    if (stripped === "tools:") {
      in_tools = true;
      current = null;
      skip_until_outdent = false;
      continue;
    }
    if (in_tools) {
      const indent = raw.length - lstrip_spaces(raw).length;
      if (skip_until_outdent && indent > 4) {
        continue;
      }
      skip_until_outdent = false;
      const m = _LIST_DASH_RE.exec(raw);
      if (m && indent === 2) {
        const first = m[1]!;
        current = {};
        tools.push(current);
        const inline = _TOP_KEY_RE.exec(first);
        if (inline) {
          current[inline[1]!] = inline[2]!;
        }
        continue;
      }
      const mk = _INDENT_KEY_RE.exec(raw);
      if (mk && current !== null && indent === 4) {
        const key = mk[1]!;
        const val = mk[2]!;
        if ((key === "files" || key === "merged_keys") && !val) {
          skip_until_outdent = true;
          continue;
        }
        current[key] = val;
        continue;
      }
    }
    const m_top = _TOP_KEY_RE.exec(raw);
    if (m_top) {
      const key = m_top[1]!;
      const value = m_top[2]!;
      if (key === "deploy_roots" && !value) {
        // Top-level v2 array — skip until next top-level scalar.
        in_tools = false;
        current = null;
        skip_until_outdent = true;
        continue;
      }
      if (key === "schema_version") {
        const parsed = parse_int_strict(value);
        data[key] = parsed === null ? value : parsed;
      } else {
        data[key] = value;
      }
      in_tools = false;
      current = null;
      skip_until_outdent = false;
    }
  }
  return data;
}

/** `str.splitlines()` — split on `\n`, dropping a trailing empty segment. */
function splitlines(text: string): string[] {
  const parts = text.split("\n");
  if (parts.length > 0 && parts[parts.length - 1] === "") {
    parts.pop();
  }
  return parts;
}

/** Left-strip only spaces, mirroring Python's `lstrip(" ")`. */
function lstrip_spaces(s: string): string {
  let i = 0;
  while (i < s.length && s[i] === " ") {
    i += 1;
  }
  return s.slice(i);
}

/** Parse a base-10 int the way Python's `int(value)` does; null on failure. */
function parse_int_strict(value: string): number | null {
  if (!/^[+-]?\d+$/.test(value.trim())) {
    return null;
  }
  return Number.parseInt(value, 10);
}

// ---------------------------------------------------------------------------
// Write
// ---------------------------------------------------------------------------

function _render(
  version: string,
  tools: AnyDict[],
  options: { deploy_roots?: string[] | null } = {},
): string {
  const deploy_roots = options.deploy_roots ?? null;
  const lines: string[] = [
    `schema_version: ${SCHEMA_VERSION}`,
    `agent_config_version: "${version}"`,
  ];
  if (deploy_roots && deploy_roots.length > 0) {
    lines.push("deploy_roots:");
    for (const root of deploy_roots) {
      lines.push(`  - ${root}`);
    }
  }
  lines.push("tools:");
  for (const tool of tools) {
    lines.push(`  - name: ${String(tool["name"])}`);
    lines.push(`    scope: ${String(tool["scope"])}`);
    lines.push(`    bridge_marker: ${String(tool["bridge_marker"])}`);
    lines.push(`    installed_at: "${String(tool["installed_at"])}"`);
    const status = tool["status"];
    if (status) {
      lines.push(`    status: ${String(status)}`);
    }
    let files = (tool["files"] as AnyDict[] | undefined) ?? [];
    if (files.length > 0) {
      // Sort by path ascending — deterministic output for golden-file tests
      // and stable team diffs (P1.3).
      files = stable_sort(files, (f) => [String(f["path"])]);
      lines.push("    files:");
      for (const entry of files) {
        lines.push(`      - path: ${String(entry["path"])}`);
        lines.push(`        kind: ${String(entry["kind"])}`);
        const sha = entry["sha256"];
        if (sha === null || sha === undefined) {
          lines.push("        sha256: null");
        } else {
          lines.push(`        sha256: "${String(sha)}"`);
        }
      }
    }
    let merged = (tool["merged_keys"] as AnyDict[] | undefined) ?? [];
    if (merged.length > 0) {
      // Sort by (file, json_pointer) ascending — deterministic output
      // regardless of insertion order (P1.3).
      merged = stable_sort(merged, (e) => [
        String(e["file"]),
        String(e["json_pointer"]),
      ]);
      lines.push("    merged_keys:");
      for (const entry of merged) {
        lines.push(`      - file: ${String(entry["file"])}`);
        lines.push(`        json_pointer: "${String(entry["json_pointer"])}"`);
        const vh = entry["value_hash"];
        if (vh !== null && vh !== undefined) {
          lines.push(`        value_hash: "${String(vh)}"`);
        }
      }
    }
  }
  return lines.join("\n") + "\n";
}

/**
 * Stable sort by a tuple key, mirroring Python's tuple-comparison `sorted`.
 * Compares each component lexicographically; ties keep input order.
 */
function stable_sort<T>(items: T[], key: (item: T) => string[]): T[] {
  return items
    .map((item, index) => ({ item, index, k: key(item) }))
    .sort((a, b) => {
      const len = Math.max(a.k.length, b.k.length);
      for (let i = 0; i < len; i += 1) {
        const av = a.k[i] ?? "";
        const bv = b.k[i] ?? "";
        if (av < bv) return -1;
        if (av > bv) return 1;
      }
      return a.index - b.index;
    })
    .map((entry) => entry.item);
}

/**
 * Atomically write the manifest; return the path written.
 *
 * Delegates to {@link write_atomic} so the crash-safety guarantees (fsync
 * file, atomic rename, fsync parent dir) are shared with every other v2
 * writer.
 *
 * `deploy_roots` is the optional top-level v2 field listing directories the
 * doctor command surveys for foreign files. When omitted, the field is not
 * emitted (callers may rely on {@link DEFAULT_DEPLOY_ROOTS} for the survey
 * scope).
 */
export function write_manifest(
  path: string,
  version: string,
  tools: AnyDict[],
  options: { deploy_roots?: string[] | null } = {},
): string {
  const rendered = _render(version, tools, { deploy_roots: options.deploy_roots ?? null });
  return write_atomic(path, rendered);
}

// ---------------------------------------------------------------------------
// Mutation helpers
// ---------------------------------------------------------------------------

/** Raised when an existing manifest entry conflicts with the new scope. */
export class ScopeMismatchError extends Error {
  readonly name_: string;
  readonly recorded_scope: string;
  readonly new_scope: string;

  constructor(name: string, recorded_scope: string, new_scope: string) {
    super(
      `tool '${name}' is committed as scope=${recorded_scope}; ` +
        `refusing to change it to scope=${new_scope} without --force`,
    );
    this.name = "ScopeMismatchError";
    this.name_ = name;
    this.recorded_scope = recorded_scope;
    this.new_scope = new_scope;
  }
}

/**
 * Return a new tools list with `name` added or refreshed.
 *
 * Idempotency rules from ADR-008 §Lifecycle:
 * - Same name, same scope → no-op (timestamp preserved).
 * - Same name, different scope → throw {@link ScopeMismatchError} unless
 *   `force=true`, in which case the entry is rewritten.
 * - New name → appended in install order (not alphabetised).
 *
 * `files` / `merged_keys` are the v2 per-tool inventories (P1.4). When
 * provided, they replace whatever was previously recorded on the entry — the
 * installer is authoritative for the set of artefacts it just wrote. When
 * `null`/omitted, existing values are preserved on the idempotent path and
 * absent on first-write.
 */
export function upsert_tool(
  existing: AnyDict[],
  options: {
    name: string;
    scope: string;
    bridge_marker: string;
    installed_at?: string | null;
    force?: boolean;
    files?: AnyDict[] | null;
    merged_keys?: AnyDict[] | null;
  },
): AnyDict[] {
  const {
    name,
    scope,
    bridge_marker,
    installed_at = null,
    force = false,
    files = null,
    merged_keys = null,
  } = options;

  if (!_VALID_SCOPES.includes(scope)) {
    throw new ValueError(`scope must be one of ${_VALID_SCOPES.join(",")}: '${scope}'`);
  }
  const stamp = installed_at || _today();

  const _build = (prior: AnyDict | null = null): AnyDict => {
    const entry: AnyDict = {
      name,
      scope,
      bridge_marker,
      installed_at: stamp,
    };
    if (files !== null) {
      entry["files"] = [...files];
    } else if (prior !== null && Array.isArray(prior["files"]) && (prior["files"] as unknown[]).length > 0) {
      entry["files"] = [...(prior["files"] as AnyDict[])];
    }
    if (merged_keys !== null) {
      entry["merged_keys"] = [...merged_keys];
    } else if (
      prior !== null &&
      Array.isArray(prior["merged_keys"]) &&
      (prior["merged_keys"] as unknown[]).length > 0
    ) {
      entry["merged_keys"] = [...(prior["merged_keys"] as AnyDict[])];
    }
    return entry;
  };

  const result: AnyDict[] = [];
  let found = false;
  for (const entry of existing) {
    if (entry["name"] === name) {
      found = true;
      const recorded = String(entry["scope"] ?? "");
      if (recorded === scope) {
        if (files === null && merged_keys === null) {
          // Idempotent no-op — preserve original entry.
          result.push(entry);
        } else {
          // Refresh inventories, preserve installed_at.
          const refreshed = _build(entry);
          refreshed["installed_at"] = entry["installed_at"] ?? stamp;
          result.push(refreshed);
        }
        continue;
      }
      if (!force) {
        throw new ScopeMismatchError(name, recorded, scope);
      }
      result.push(_build(entry));
      continue;
    }
    result.push(entry);
  }
  if (!found) {
    result.push(_build());
  }
  return result;
}

/** Thrown for invalid-argument conditions, mirroring Python's `ValueError`. */
export class ValueError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ValueError";
  }
}

function _today(): string {
  // UTC date in YYYY-MM-DD — matches Python's
  // `datetime.now(timezone.utc).strftime("%Y-%m-%d")`.
  return new Date().toISOString().slice(0, 10);
}
